### ISSUE TITLE
rustc: inference for numeric literals does not take default type parameters into account

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#15760]
The following code fails to compile:

``` rust
#![feature(default_type_params)]

fn foo<T = u64>(t: T) {}
fn main() {
    foo(0);
}
```

with

> error: cannot determine a type for this expression: cannot determine the type of this integer; add a suffix to specify the type explicitly

despite the fact that the actual type is determinant.

Fixing this problem could help mitigate problems with the `int` fallback: for functions like `range`, we could provide a default type parameter of, say, `uint` while retaining flexibility for other types. (Doing so loses some of the benefit of the fallback removal, but we can introduce defaults in a controlled way.)

**Comment by steveklabnik** [comment]
This code now compiles.

**Comment by eddyb** [comment]
It only compiles because we default to `i32` - presumably you want to use `size_of_val` or something similar to check the type.

**Comment by bltavares** [comment]
Triage: the OP's code compiles fine on nightly with many warnings:

```
warning: this feature is stable. attribute no longer needed, #[warn(stable_features)] on by default
 --> <anon>:1:12
1 |> #![feature(default_type_params)]
  |>            ^^^^^^^^^^^^^^^^^^^

warning: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions., #[warn(invalid_type_param_default)] on by default
 --> <anon>:3:8
3 |> fn foo<T = u64>(t: T) {}
  |>        ^
warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
note: for more information, see PR 30742 <https://github.com/rust-lang/rust/pull/30724>

warning: unused variable: `t`, #[warn(unused_variables)] on by default
 --> <anon>:3:17
3 |> fn foo<T = u64>(t: T) {}
  |>                 ^
```

On stable, without the feature pragma, the warning on default for functions will be a hard error in future release, as mentioned on the nightly warnings.

**Comment by leodasvacas** [comment]
@Mark-Simulacrum I think this issue can be closed as a subissue of #27336, the fate of this will be decided by what we do with that RFC. It's certainly not "just a bug".

**Comment by Mark-Simulacrum** [comment]
Yeah, that's true. Thanks -- closing.
